python/runtime/cudaq/algorithms/py_run.cpp

```diff
@@ -122,10 +125,108 @@ std::vector<py::object> pyRun(py::object &kernel, py::args args,
   return results;
 }
 
+// Internal struct representing buffer to be filled asynchronously.
+// When the `ready` future is set, the content of the buffer is filled.
+struct async_run_result {
+  std::future<void> ready;
+  std::vector<py::object> *results;
+  std::string *error;
+};
+
+/// @brief Run `cudaq::run_async` on the provided kernel.
+async_run_result pyRunAsync(py::object &kernel, py::args args,
+                            std::size_t shots_count,
+                            std::optional<noise_model> noise_model,
+                            std::size_t qpu_id) {
+  kernel.inc_ref();
+  auto &platform = get_platform();
+  auto numQPUs = platform.num_qpus();
+  if (qpu_id >= numQPUs)
+    throw std::runtime_error("qpu_id (" + std::to_string(qpu_id) +
+                             ") exceeds the number of available QPUs (" +
+                             std::to_string(numQPUs) + ")");
+
+  auto [name, module, argData, func] =
+      details::getKernelLaunchParameters(kernel, args);
+
+  auto mod = unwrap(module);
+  mod->setAttr(runtime::enableCudaqRun, mlir::UnitAttr::get(mod->getContext()));
+
+  if (noise_model.has_value() && platform.is_remote())
+    throw std::runtime_error(
+        "Noise model is not supported on remote platforms.");
+
+  async_run_result result;
+  std::promise<void> promise;
+  result.ready = promise.get_future();
+  result.results = new std::vector<py::object>();
+  result.error = new std::string();
+
+  if (shots_count == 0) {
+    promise.set_value();
+    return result;
+  }
+
+  // Release GIL to allow c++ threads.
+  py::gil_scoped_release gil_release{};
+
+  QuantumTask wrapped = detail::make_copyable_function(
+      [p = std::move(promise), resultPtr = result.results,
+       errorPtr = result.error, shots_count, argData, name, module, func,
+       noise_model = std::move(noise_model)]() mutable {
+        auto &platform = get_platform();
+
+        // Launch the kernel in the appropriate context.
+        if (noise_model.has_value())
+          platform.set_noise(&noise_model.value());
+
+        py::gil_scoped_acquire gil{};
```

Comment:
I wonder whether this GIL acquire would cause all the kernel executions (on different GPU-emulated QPUs) sequential.
We'd want these `QuantumTask` functors to run independently on threads.

Reply:
Yeah I think you are right. I have to hold the GIL for converting results to python objects, but I can try splitting out the c++ part out. Then only the conversion will need to be synchronized on…

Reply:
Updated to only acquire the GIL for the translation of the c++ return values to python object (it is required there)